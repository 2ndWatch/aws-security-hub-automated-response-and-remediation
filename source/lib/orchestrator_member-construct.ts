#!/usr/bin/env node
/*****************************************************************************
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

// Create a new stack to deploy a Playbook in a member account
// Orchestrator member role
// Remediation role(s)
// SSM Automation Documents

import * as cdk from '@aws-cdk/core';
import { 
    PolicyStatement, 
    Effect, 
    Role, 
    PolicyDocument, 
    ArnPrincipal, 
    CompositePrincipal, 
    CfnRole 
} from '@aws-cdk/aws-iam';

export interface SHARRStackProps extends cdk.StackProps  {
    solutionId: string;
    solutionVersion: string;
    solutionDistBucket: string;
    solutionDistName: string;
    solutionName: string;
    securityStandard: string;
}

export class MemberStack extends cdk.Stack {

  SEND_ANONYMOUS_DATA = 'Yes'
  public readonly adminAccountNumber: cdk.CfnParameter;

  constructor(scope: cdk.App, id: string, props: SHARRStackProps) {
    super(scope, id, props);

    const validAwsAccount = "\\d{12}"
    const adminRoleName = props.solutionId + '-SHARR-Orchestrator-Admin_' + this.region

    this.adminAccountNumber = new cdk.CfnParameter(this, 'Admin Account Number', {
        description: "Admin account number",
        type: "String",
        allowedPattern: validAwsAccount
    });

    const memberPolicy = new PolicyDocument();
    /**
     * @description Cross-account permissions for Orchestration role
     * @type {PolicyStatement}
     */
    const ssmROPerms = new PolicyStatement()
    ssmROPerms.addActions("ssm:DescribeAutomationExecutions")
    ssmROPerms.addActions("ssm:DescribeDocument")
    ssmROPerms.effect = Effect.ALLOW;
    ssmROPerms.addResources(
        "arn:" + this.partition + ":ssm:" + this.region + ":*:*"
    )
    memberPolicy.addStatements(ssmROPerms)

    let principalPolicyStatement = new PolicyStatement();

    principalPolicyStatement.addActions("sts:AssumeRole");
    principalPolicyStatement.effect = Effect.ALLOW;

    let roleprincipal = new ArnPrincipal(
        'arn:' + this.partition + ':iam::' + this.adminAccountNumber.value + ':role/' +
        adminRoleName
    );

    let principals = new CompositePrincipal(roleprincipal);
    principals.addToPolicy(principalPolicyStatement);

    let memberRole = new Role(this, 'MemberAccountRole', {
        assumedBy: principals,
        inlinePolicies: {
            'default_lambdaPolicy': memberPolicy
        },
        roleName: props.solutionId + '-SHARR-Orchestrator-Member-' + props.securityStandard + '_' + this.region
    });

    const memberRoleResource = memberRole.node.findChild('Resource') as CfnRole;

    memberRoleResource.cfnOptions.metadata = {
        cfn_nag: {
            rules_to_suppress: [{
                id: 'W11',
                reason: 'Resource * is required due to the administrative nature of the solution.'
            },{
                id: 'W28',
                reason: 'Static names chosen intentionally to provide integration in cross-account permissions'
            }]
        }
    };
  }
}


