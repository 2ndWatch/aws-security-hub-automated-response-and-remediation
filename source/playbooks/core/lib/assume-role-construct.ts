/*****************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import { PolicyStatement, Effect, Role, PolicyDocument, ArnPrincipal, ServicePrincipal, CompositePrincipal, CfnRole } from '@aws-cdk/aws-iam';

export interface AssumeRoleConstructProps {
    masterAccountNumber: cdk.CfnParameter,
    solutionId: string,
    lambdaPolicy: PolicyDocument,
    lambdaHandlerName: string
    service?: string
}

export class AssumeRoleConstruct extends cdk.Construct {

    constructor(scope: cdk.Construct, id: string, props: AssumeRoleConstructProps) {
        super(scope, id);

        //DEFINE PRINCIPLE RESOURCE WHICH WILL ASSUME THE ROLE.

        let principalPolicyStatement = new PolicyStatement();
        principalPolicyStatement.addActions("sts:AssumeRole");
        principalPolicyStatement.effect = Effect.ALLOW;

        let roleprincipal = new ArnPrincipal('arn:aws:iam::' + props.masterAccountNumber.value + ':role/' +
            props.solutionId + '_' + props.lambdaHandlerName +  '_lambdaRole');

        let principals = new CompositePrincipal(roleprincipal);
        principals.addToPolicy(principalPolicyStatement);

        if (props.service) {
            let serviceprincipal = new ServicePrincipal(props.service)
            principals.addPrincipals(serviceprincipal);
        }

        let lambdaRole = new Role(this, 'MemberAccountRole', {
            assumedBy: principals,
            inlinePolicies: {
                'default_lambdaPolicy': props.lambdaPolicy
            },
            roleName: props.solutionId + '_' + props.lambdaHandlerName + '_memberRole'
        });

        const lambdaRoleResource = lambdaRole.node.findChild('Resource') as CfnRole;

        lambdaRoleResource.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W11',
                    reason: 'Resource * is required due to the administrative nature of the solution.'
                },{
                    id: 'W28',
                    reason: 'Static names chosen intentionally to provide integration in cross-account permissions'
                }]
            }
        };
    }
}