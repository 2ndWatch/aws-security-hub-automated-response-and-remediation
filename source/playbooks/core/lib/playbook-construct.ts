/*****************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as events from '@aws-cdk/aws-events';
import * as eventstarget from '@aws-cdk/aws-events-targets';
import {Effect, PolicyDocument, PolicyStatement, ServicePrincipal, Policy, Role, CfnRole} from '@aws-cdk/aws-iam';
import { Bucket, IBucket } from '@aws-cdk/aws-s3';
import * as e2l from '@aws-solutions-constructs/aws-events-rule-lambda';

export interface IPlaybookConstructProps {
    name: string;
    description: string;
    lambda_env?: any;
    aws_region: string;
    aws_accountid: string;
    lambda_handler?: string;
    lambda_memsize?: number;
    lambda_maxtime?: number;
    custom_action_name: string;
    findings: object;
    solutionId: string;
    solutionVersion: string;
    solutionName: string;
    distName: string;
    distBucket: string;
}

export class PlaybookConstruct extends cdk.Construct {
    
    constructor(scope: cdk.Construct, id: string, props: IPlaybookConstructProps) {
        super(scope, id);
        let workflowStatusFilter = {
            "Status": [ "NEW" ]
        }
        let complianceStatusFilter = {
            "Status": [ "FAILED", "WARNING" ]
        }

        let RESOURCE_PREFIX = props.solutionId;

        let lambdaRole = undefined

        const basePolicy = new PolicyStatement();
        basePolicy.addActions("cloudwatch:PutMetricData")
        basePolicy.addActions("securityhub:BatchUpdateFindings")
        basePolicy.effect = Effect.ALLOW
        basePolicy.addResources("*")

        const logsPolicy = new PolicyStatement();
        logsPolicy.addActions("logs:CreateLogGroup")
        logsPolicy.addActions("logs:CreateLogStream")
        logsPolicy.addActions("logs:PutLogEvents")
        logsPolicy.effect = Effect.ALLOW
        logsPolicy.addResources("*")

        const kmsPolicy = new PolicyStatement();
        kmsPolicy.addActions("kms:Encrypt")
        kmsPolicy.addActions("kms:Decrypt")
        kmsPolicy.addActions("kms:GenerateDataKey")
        kmsPolicy.effect = Effect.ALLOW
        kmsPolicy.addResources('{{resolve:ssm:/Solutions/' + props.solutionId + '/CMK_ARN:1}}')

        const snsPolicy = new PolicyStatement();
        snsPolicy.addActions("sns:Publish")
        snsPolicy.effect = Effect.ALLOW
        snsPolicy.addResources('arn:aws:sns:' + props.aws_region + ':' +
            props.aws_accountid + ':' + props.solutionId + '-SHARR_Topic')

        const stsPolicy = new PolicyStatement();
        stsPolicy.addActions("sts:AssumeRole")
        stsPolicy.effect = Effect.ALLOW
        stsPolicy.addResources('arn:aws:iam::*:role/' +
            RESOURCE_PREFIX + '_' + props.name + '_memberRole_' + props.aws_region)

        const lambdaPolicy = new PolicyDocument();
        lambdaPolicy.addStatements(logsPolicy)
        lambdaPolicy.addStatements(basePolicy)
        lambdaPolicy.addStatements(snsPolicy)
        lambdaPolicy.addStatements(kmsPolicy)
        lambdaPolicy.addStatements(stsPolicy)

        const principal = new ServicePrincipal('lambda.amazonaws.com');
        const principalPolicyStatement = new PolicyStatement();
        principalPolicyStatement.addActions("sts:AssumeRole");
        principalPolicyStatement.effect = Effect.ALLOW;
        principal.addToPolicy(principalPolicyStatement);

        let roleName: string = RESOURCE_PREFIX + '_' + props.name + '_lambdaRole_' + props.aws_region;
        lambdaRole = new Role(this, 'Role', {
            assumedBy: principal,
            inlinePolicies: {
                'default_lambdaPolicy': lambdaPolicy
            },
            roleName: roleName
        });

        const lambdaRoleResource = lambdaRole.node.findChild('Resource') as CfnRole;

        lambdaRoleResource.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W11',
                    reason: 'Resource * is required due to the administrative nature of the solution.'
                },{
                    id: 'W28',
                    reason: 'Static names chosen intentionally to provide integration in cross-account permissions'
                }]
            }
        };

        let lambdaName: string = props.name + '_lambda';

        let s3BucketForLambda: IBucket = Bucket.fromBucketAttributes(this, "s3BucketForLambda", {
            bucketName: props.distBucket + '-' + props.aws_region
        })

        let lambda_src = props.distName + '/'
            + props.solutionVersion + '/playbooks/CIS/' 
            + props.name.toLowerCase() + '.py.zip';

        let lambda_handler = undefined
        if (props.lambda_handler) {
            lambda_handler = props.lambda_handler
        } else {
            lambda_handler = props.name.toLowerCase() + '.lambda_handler';
        }

        let lambda_memsize = 256
        if (props.lambda_memsize) {
            lambda_memsize = props.lambda_memsize
        }

        let lambda_maxtime = 60
        if (props.lambda_maxtime) {
            lambda_maxtime = props.lambda_maxtime
        }

        // Event to Lambda
        // ---------------
        let eventRuleName: string = props.name + '_eventRule'

        let customActionName: string = props.name + '_customAction';
        const customAction = new cdk.CustomResource(this, 'CustomAction', {
            serviceToken: 'arn:aws:lambda:' + props.aws_region + ':' + 
                props.aws_accountid + ':function:' + RESOURCE_PREFIX + '-SHARR-CustomAction',
            resourceType: 'Custom::ActionTarget',
            properties: {
                Name: props.custom_action_name,
                Description: props.description,
                Id: props.name,
            }
        });

        const e2lprops = {
            deployLambda: true,
            lambdaFunctionProps: {
                functionName: lambdaName,
                description: 'SHARR-' + props.description,
                runtime: lambda.Runtime.PYTHON_3_8,
                code: lambda.Code.fromBucket(s3BucketForLambda, lambda_src),
                handler: lambda_handler,
                memorySize: lambda_memsize,
                role: lambdaRole,
                timeout: cdk.Duration.seconds(lambda_maxtime),
                environment: {
                    log_level: 'info',
                    sendAnonymousMetrics: '{{resolve:ssm:/Solutions/' + props.solutionId + '/sendAnonymousMetrics:1}}',
                    metricsId: '{{resolve:ssm:/Solutions/' + props.solutionId + '/metricsId:1}}'
                }
            },
            eventRuleProps: {
                description: props.description + ' event rule.',
                ruleName: eventRuleName,
                enabled: true,
                eventPattern: {
                    source: ["aws.securityhub"],
                    detailType: ["Security Hub Findings - Custom Action"],
                    resources: [customAction.getAttString('Arn')],
                    detail: {
                        findings: { 
                            Title: props.findings,
                            Workflow: workflowStatusFilter,
                            Compliance: complianceStatusFilter
                        }
                    }
                }
            }
        }

        let eventToLambda = new e2l.EventsRuleToLambda(this, 'eventToPlaybook', e2lprops)

        if (props.lambda_env) {
            let envs = Object.keys(props.lambda_env);
            for (let env of envs) {
                eventToLambda.lambdaFunction.addEnvironment(env, props.lambda_env[env])
            }
        }

        const eventTarget = new eventstarget.LambdaFunction(eventToLambda.lambdaFunction);

        const enable_auto_remediation_param = new cdk.CfnParameter(this, 'AutoEnable', {
            description: "This will fully enable automated remediation for "+ props.description.slice(11),
            type: "String",
            allowedValues: ["ENABLED", "DISABLED"],
            default: "DISABLED"
        });

        enable_auto_remediation_param.overrideLogicalId(props.name + "AutoRemediation")

        // Adding an automated even rule for the playbook
        const eventRule_auto = new events.Rule(this, 'AutoEventRule', {
            description: props.description + ' automatic remediation event rule.',
            ruleName: eventRuleName + '_automated',
            targets: [eventTarget]
        });

        const cfnEventRule_auto = eventRule_auto.node.defaultChild as events.CfnRule;
        cfnEventRule_auto.addPropertyOverride('State', enable_auto_remediation_param.valueAsString);

        eventRule_auto.addEventPattern({
            source: ["aws.securityhub"],
            detailType: ["Security Hub Findings - Imported"],
            detail: {
                findings: { 
                    Title: props.findings,
                    Workflow: workflowStatusFilter,
                    Compliance: complianceStatusFilter
                }
            }
        });
    }
}