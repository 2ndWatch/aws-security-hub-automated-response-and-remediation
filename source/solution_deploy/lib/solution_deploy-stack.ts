/*****************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as logs from '@aws-cdk/aws-logs';
import * as sc from '@aws-cdk/aws-servicecatalog';
import * as iam from '@aws-cdk/aws-iam';
import * as s3 from '@aws-cdk/aws-s3';
import * as sns from '@aws-cdk/aws-sns';
import * as lambda from '@aws-cdk/aws-lambda';
import * as ssm from '@aws-cdk/aws-ssm';
import * as kms from '@aws-cdk/aws-kms';
import * as fs from 'fs';

export interface SHARRStackProps extends cdk.StackProps  {
    solutionId: string;
    solutionVersion: string;
    solutionDistBucket: string;
    solutionTMN: string;
    solutionName: string;
}

export class SolutionDeployStack extends cdk.Stack {

  SEND_ANONYMOUS_DATA = 'Yes'

  constructor(scope: cdk.App, id: string, props: SHARRStackProps) {
    super(scope, id, props);
    
    const RESOURCE_PREFIX = props.solutionId; // prefix on every resource name

    //-------------------------------------------------------------------------
    // Solutions Bucket - Source Code
    //
    const SolutionsBucket = s3.Bucket.fromBucketAttributes(this, 'SolutionsBucket', {
        bucketName: props.solutionDistBucket + '-' + this.region
    });

    //=========================================================================
    // MAPPINGS
    //=========================================================================
    new cdk.CfnMapping(this, 'SourceCode', {
        mapping: { "General": { 
            "S3Bucket": props.solutionDistBucket,
            "KeyPrefix": props.solutionTMN + '/' + props.solutionVersion
        } }
    })

    //-------------------------------------------------------------------------
    // KMS Key for solution encryption
    //
    const kmsKey = new kms.Key(this, 'SHARR-key', {
        enableKeyRotation: true,
        alias: RESOURCE_PREFIX + '-SHARR-Key',
        trustAccountIdentities: true
    });

    new ssm.StringParameter(this, 'SHARR_Key', {
        description: 'KMS Customer Managed Key that SHARR will use to encrypt data',
        parameterName: '/Solutions/' + RESOURCE_PREFIX + '/CMK_ARN',
        stringValue: kmsKey.keyArn
    });

    //-------------------------------------------------------------------------
    // SNS Topic for notification fanout on Playbook completion
    //
    const snsTopic = new sns.Topic(this, 'SHARR-Topic', {
        displayName: 'SHARR Playbook Topic (' + RESOURCE_PREFIX + ')',
        topicName: RESOURCE_PREFIX + '-SHARR_Topic',
        masterKey: kmsKey
    });

    new ssm.StringParameter(this, 'SHARR_SNS_Topic', {
        description: 'SNS Topic ARN where SHARR will send status messages. This\
        topic can be useful for driving additional actions, such as email notifications,\
        trouble ticket updates.',
        parameterName: '/Solutions/' + RESOURCE_PREFIX + '/SNS_Topic_ARN',
        stringValue: snsTopic.topicArn
    });

    const mapping = new cdk.CfnMapping(this, 'mappings');
    mapping.setValue("sendAnonymousMetrics", "data", this.SEND_ANONYMOUS_DATA)

	new ssm.StringParameter(this, 'SHARR_SendAnonymousMetrics', {
		description: 'Flag to enable or disable sending anonymous metrics.',
		parameterName: '/Solutions/' + RESOURCE_PREFIX + '/sendAnonymousMetrics',
		stringValue: mapping.findInMap("sendAnonymousMetrics", "data")
	});

	new ssm.StringParameter(this, 'SHARR_metricsId', {
		description: 'Stack Id for reference for metrics.',
		parameterName: '/Solutions/' + RESOURCE_PREFIX + '/metricsId',
		stringValue: this.stackId
	});

    //-------------------------------------------------------------------------
    // Custom Lambda Policy
    //
    const createCustomActionPolicy = new iam.Policy(this, 'createCustomActionPolicy', {
        policyName: RESOURCE_PREFIX + '-SHARR_Custom_Action',
        statements: [
            new iam.PolicyStatement({
                actions: [
                    'cloudwatch:PutMetricData'
                ],
                resources: ['*']
            }),
            new iam.PolicyStatement({
                actions: [
                    'logs:CreateLogGroup',
                    'logs:CreateLogStream',
                    'logs:PutLogEvents'
                ],
                resources: ['*']
            }),
            new iam.PolicyStatement({
                actions: [
                    'securityhub:CreateActionTarget',
                    'securityhub:DeleteActionTarget'
                ],
                resources: ['*']
            })
        ]
    })

    const createCAPolicyResource = createCustomActionPolicy.node.findChild('Resource') as iam.CfnPolicy;

    createCAPolicyResource.cfnOptions.metadata = {
        cfn_nag: {
            rules_to_suppress: [{
                id: 'W12',
                reason: 'Resource * is required for CloudWatch Logs policies used on Lambda functions.'
            }]
        }
    };

    //-------------------------------------------------------------------------
    // Custom Lambda Role
    //
    const createCustomActionRole = new iam.Role(this, 'createCustomActionRole', {
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        description: 'Lambda role to allow creation of Security Hub Custom Actions'
    });

    createCustomActionRole.attachInlinePolicy(createCustomActionPolicy);

    const createCARoleResource = createCustomActionRole.node.findChild('Resource') as iam.CfnRole;

    createCARoleResource.cfnOptions.metadata = {
        cfn_nag: {
            rules_to_suppress: [{
                id: 'W28',
                reason: 'Static names chosen intentionally to provide easy integration with playbook templates'
            }]
        }
    };

    //-------------------------------------------------------------------------
    // Custom Lambda - Create Custom Action
    //
    const createCustomAction = new lambda.Function(this, 'CreateCustomAction', {
        functionName: RESOURCE_PREFIX + '-SHARR-CustomAction',
        handler: 'createCustomAction.lambda_handler',
        runtime: lambda.Runtime.PYTHON_3_8,
        description: 'Custom resource to create an action target in Security Hub',
        code: lambda.Code.fromBucket(
            SolutionsBucket,
            props.solutionTMN + '/' + props.solutionVersion + '/lambda/createCustomAction.py.zip'
        ),
        environment: {
            log_level: 'info',
            sendAnonymousMetrics: mapping.findInMap("sendAnonymousMetrics", "data"),
            AWS_PARTITION: this.partition
        },
        memorySize: 256,
        timeout: cdk.Duration.seconds(60),
        role: createCustomActionRole
    });

    const createCAFuncResource = createCustomAction.node.findChild('Resource') as lambda.CfnFunction;

    createCAFuncResource.cfnOptions.metadata = {
        cfn_nag: {
            rules_to_suppress: [{
                id: 'W58',
                reason: 'False positive. See https://github.com/stelligent/cfn_nag/issues/422'
            }]
        }
    };

    //---------------------------------------------------------------------
    // Service Catalog Nested Stack
    //
    const serviceCatalog = new cdk.CfnStack(this, "PlaybookServiceCatalog", {
        parameters: {
            CreateCustomActionArn: createCustomAction.functionArn
        },
        templateUrl: "https://" + cdk.Fn.findInMap("SourceCode", "General", "S3Bucket") +
        "-reference.s3.amazonaws.com/" + cdk.Fn.findInMap("SourceCode", "General", "KeyPrefix") +
        "/aws-sharr-portolio-deploy.template"
    })
    serviceCatalog.cfnOptions.condition = new cdk.CfnCondition(this, "UseServiceCatalog", {
        expression: cdk.Fn.conditionNot(cdk.Fn.conditionEquals(this.partition, "aws-cn"))
    });
  }
}
