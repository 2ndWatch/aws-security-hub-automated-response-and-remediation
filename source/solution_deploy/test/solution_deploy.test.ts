import { expect as expectCDK, matchTemplate, MatchStyle, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as SolutionDeploy from '../lib/solution_deploy-stack';

function getTestStack(): cdk.Stack {
  const envEU = { account: '111111111111', region: 'eu-west-1' };
  const app = new cdk.App();
  const stack = new SolutionDeploy.SolutionDeployStack(app, 'stack', { env: envEU })
  return stack;
}

test('Test if the Stack has all the resources.', () => {
  process.env.DIST_OUTPUT_BUCKET = 'solutions'
  process.env.SOLUTION_NAME = 'AWS Security Hub Automated Response & Remediation'
  process.env.DIST_VERSION = 'v1.0.0'
  process.env.SOLUTION_ID = 'SO0111'
  process.env.SOLUTION_TRADEMARKEDNAME = 'aws-security-hub-automated-response-and-remediation'
  expect(SynthUtils.toCloudFormation(getTestStack())).toMatchSnapshot();
});
