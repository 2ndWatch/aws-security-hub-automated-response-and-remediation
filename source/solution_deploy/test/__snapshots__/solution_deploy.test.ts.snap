// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test if the Stack has all the resources. 1`] = `
Object {
  "Conditions": Object {
    "UseServiceCatalog": Object {
      "Fn::Not": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Partition",
            },
            "aws-cn",
          ],
        },
      ],
    },
  },
  "Mappings": Object {
    "SourceCode": Object {
      "General": Object {
        "KeyPrefix": "aws-security-hub-automated-response-and-remediation/v1.0.0",
        "S3Bucket": "solutions",
      },
    },
    "mappings": Object {
      "sendAnonymousMetrics": Object {
        "data": "Yes",
      },
    },
  },
  "Resources": Object {
    "CreateCustomActionE7A973F5": Object {
      "DependsOn": Array [
        "createCustomActionRoleF0047414",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "False positive. See https://github.com/stelligent/cfn_nag/issues/422",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "solutions-eu-west-1",
          "S3Key": "aws-security-hub-automated-response-and-remediation/v1.0.0/lambda/createCustomAction.py.zip",
        },
        "Description": "Custom resource to create an action target in Security Hub",
        "Environment": Object {
          "Variables": Object {
            "AWS_PARTITION": Object {
              "Ref": "AWS::Partition",
            },
            "log_level": "info",
            "sendAnonymousMetrics": Object {
              "Fn::FindInMap": Array [
                "mappings",
                "sendAnonymousMetrics",
                "data",
              ],
            },
          },
        },
        "FunctionName": "SO0111-SHARR-CustomAction",
        "Handler": "createCustomAction.lambda_handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "createCustomActionRoleF0047414",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "PlaybookServiceCatalog": Object {
      "Condition": "UseServiceCatalog",
      "Properties": Object {
        "Parameters": Object {
          "CreateCustomActionArn": Object {
            "Fn::GetAtt": Array [
              "CreateCustomActionE7A973F5",
              "Arn",
            ],
          },
        },
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://",
              Object {
                "Fn::FindInMap": Array [
                  "SourceCode",
                  "General",
                  "S3Bucket",
                ],
              },
              "-reference.s3.amazonaws.com/",
              Object {
                "Fn::FindInMap": Array [
                  "SourceCode",
                  "General",
                  "KeyPrefix",
                ],
              },
              "/aws-sharr-portolio-deploy.template",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
    },
    "SHARRKeyC551FE02": Object {
      "Properties": Object {
        "Description": "KMS Customer Managed Key that SHARR will use to encrypt data",
        "Name": "/Solutions/SO0111/CMK_ARN",
        "Type": "String",
        "Value": Object {
          "Fn::GetAtt": Array [
            "SHARRkeyE6BD0F56",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SHARRSNSTopicB940F479": Object {
      "Properties": Object {
        "Description": "SNS Topic ARN where SHARR will send status messages. This        topic can be useful for driving additional actions, such as email notifications,        trouble ticket updates.",
        "Name": "/Solutions/SO0111/SNS_Topic_ARN",
        "Type": "String",
        "Value": Object {
          "Ref": "SHARRTopic229CFB9E",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SHARRSendAnonymousMetricsCDAE439D": Object {
      "Properties": Object {
        "Description": "Flag to enable or disable sending anonymous metrics.",
        "Name": "/Solutions/SO0111/sendAnonymousMetrics",
        "Type": "String",
        "Value": Object {
          "Fn::FindInMap": Array [
            "mappings",
            "sendAnonymousMetrics",
            "data",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SHARRTopic229CFB9E": Object {
      "Properties": Object {
        "DisplayName": "SHARR Playbook Topic (SO0111)",
        "KmsMasterKeyId": Object {
          "Fn::GetAtt": Array [
            "SHARRkeyE6BD0F56",
            "Arn",
          ],
        },
        "TopicName": "SO0111-SHARR_Topic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "SHARRkeyAlias37E34763": Object {
      "Properties": Object {
        "AliasName": "alias/SO0111-SHARR-Key",
        "TargetKeyId": Object {
          "Fn::GetAtt": Array [
            "SHARRkeyE6BD0F56",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "SHARRkeyE6BD0F56": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "EnableKeyRotation": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::111111111111:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "SHARRmetricsId08072E2E": Object {
      "Properties": Object {
        "Description": "Stack Id for reference for metrics.",
        "Name": "/Solutions/SO0111/metricsId",
        "Type": "String",
        "Value": Object {
          "Ref": "AWS::StackId",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "createCustomActionPolicyE424E925": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Resource * is required for CloudWatch Logs policies used on Lambda functions.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "securityhub:CreateActionTarget",
                "securityhub:DeleteActionTarget",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SO0111-SHARR_Custom_Action",
        "Roles": Array [
          Object {
            "Ref": "createCustomActionRoleF0047414",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "createCustomActionRoleF0047414": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W28",
              "reason": "Static names chosen intentionally to provide easy integration with playbook templates",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Lambda role to allow creation of Security Hub Custom Actions",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
